src/dragonball/src/device_manager/console_manager.rs

```diff
@@ -0,0 +1,427 @@
+// Copyright (C) 2022 Alibaba Cloud. All rights reserved.
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+//! Virtual machine console device manager.
+//!
+//! A virtual console are composed up of two parts: frontend in virtual machine and backend in
+//! host OS. A frontend may be serial port, virtio-console etc, a backend may be stdio or Unix
+//! domain socket. The manager connects the frontend with the backend.
+//!
+//! Note: unwrap() is used without documentation, it's OK here.
+use std::io::{self, Read};
+use std::os::unix::net::{UnixListener, UnixStream};
+use std::path::Path;
+use std::sync::{Arc, Mutex};
+
+use bytes::{BufMut, BytesMut};
+use dbs_legacy_devices::{ConsoleHandler, SerialDevice};
+use dbs_utils::epoll_manager::{
+    EpollManager, EventOps, EventSet, Events, MutEventSubscriber, SubscriberId,
+};
+use vmm_sys_util::terminal::Terminal;
+
+use super::{DeviceMgrError, Result};
+
+const EPOLL_EVENT_SERIAL: u32 = 0;
+const EPOLL_EVENT_SERIAL_DATA: u32 = 1;
+const EPOLL_EVENT_STDIN: u32 = 2;
+
+/// Errors related to Console manager operations.
+#[derive(Debug, thiserror::Error)]
+pub enum ConsoleManagerError {
+    /// Cannot create unix domain socket for serial port
+    #[error("cannot create socket for serial console")]
+    CreateSerialSock(#[source] std::io::Error),
+
+    /// An operation on the epoll instance failed due to resource exhaustion or bad configuration.
+    #[error("failure while managing epoll event for console fd")]
+    EpollMgr(#[source] dbs_utils::epoll_manager::Error),
+
+    /// Cannot set mode for terminal.
+    #[error("failure while setting attribute for terminal")]
+    StdinHandle(#[source] vmm_sys_util::errno::Error),
+}
+
+/// Console manager to manage frontend and backend console devices.
+pub struct ConsoleManager {
+    epoll_mgr: EpollManager,
+    logger: slog::Logger,
+    subscriber_id: Option<SubscriberId>,
+    stdin_handle: Option<std::io::Stdin>,
+    sock_path: Option<String>,
```

Comment:
It appears (?) that a console manager must either set the `stdin_handle` or `sock_path`, but not both. If true, This could be an enum containing either the `std::io::Stdin` or the socket path `String`. That would make the code clearer I think.

Reply:
Yes, I have changed that to an enum.